js/packages/store/provider.tsx

```diff
@@ -1,365 +1,106 @@
-				if (precancel) {
-					return
-				}
-				stream.onMessage((msg, err) => {
-					if (err) {
-						console.warn('events stream onMessage error:', err)
-						dispatch({ type: 'SET_STREAM_ERROR', payload: { error: err } })
-						return
-					}
-					const evt = msg && msg.event
-					if (!evt) {
-						console.warn('received empty event')
-						return
-					}
-					const enumName = Object.keys(messengerpb.StreamEvent.Type).find(
-						(name) => messengerpb.StreamEvent.Type[name] === evt.type,
-					)
-					const payloadName = enumName.substr('Type'.length)
-					const pbobj = messengerpb.StreamEvent[payloadName]
-					if (!pbobj) {
-						console.warn('failed to find a protobuf object matching the event type')
-						return
-					}
-					const eventPayload = pbobj.decode(evt.payload)
-					if (evt.type === messengerpb.StreamEvent.Type.TypeNotified) {
-						const enumName = Object.keys(messengerpb.StreamEvent.Notified.Type).find(
-							(name) => messengerpb.StreamEvent.Notified.Type[name] === eventPayload.type,
-						)
-						const payloadName = enumName.substr('Type'.length)
-						const pbobj = messengerpb.StreamEvent.Notified[payloadName]
-						if (!pbobj) {
-							console.warn('failed to find a protobuf object matching the notification type')
-							return
-						}
-						eventPayload.payload = pbobj.decode(eventPayload.payload)
-						eventEmitter.emit('notification', {
-							type: eventPayload.type,
-							name: payloadName,
-							payload: eventPayload,
-						})
-					} else {
-						dispatch({
-							type: evt.type,
-							name: payloadName,
-							payload: eventPayload,
-						})
-					}
-				})
-				cancel = await stream.start()
-			})
-			.catch((err) => {
-				if (err === EOF) {
-					console.info('end of the events stream')
-				} else {
-					console.warn('events stream error:', err)
-				}
-				dispatch({ type: 'SET_STREAM_ERROR', payload: { error: err } })
-			})
-		return () => cancel()
-	}, [embedded, nodeStarted, state.daemonAddress, eventEmitter])
-
-	React.useEffect(() => {
-		if (!state.convsClosed && state.listDone) {
-			dispatch({ type: 'CONVS_CLOSED' })
-			for (const conv of Object.values(state.conversations) as any) {
-				if (conv.isOpen) {
-					state.client.conversationClose({ groupPk: conv.publicKey }).catch((e: any) => {
-						console.warn(`failed to close conversation "${conv.displayName}",`, e)
-					})
-				}
-			}
-		}
-	}, [state.client, state.conversations, state.convsClosed, state.listDone])
-
 	return (
 		<MsgrContext.Provider
 			value={{
 				...state,
-				restart,
-				deleteAccount,
 				dispatch,
-				addNotificationListener,
-				removeNotificationListener,
-				setPersistentOption,
+				restart: () => dispatch({ type: MessengerActions.Restart }),
+				deleteAccount: () => dispatch({ type: MessengerActions.SetStateDeleting }),
+				switchToAccount: (accountID: number) =>
+					dispatch({ type: MessengerActions.SetNextAccount, payload: accountID }),
```

Comment:
all these really need to be wrapped in useCallback, `dispatch` never changes, so an effect depending on it will not be retriggered, here the `restart` function will be recreated every time the app state changes

Reply:
```js
const useRestart = () => {
  const ctx = useMsgrContext()
  return useCallback(() => ctx.dispatch({ type: MessengerActions.Restart }), [ctx.dispatch])
}
```